### PR TITLE
Bulk update of external shard allocation

### PR DESCRIPTION
**Pull request by chbatey**
References #28696

### CODE REVIEW

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/internal/ExternalShardAllocationClientImpl.scala [hunk 1]
```diff
@@ -91,4 +91,22 @@ final private[external] class ExternalShardAllocationClientImpl(system: ActorSys
   }
 
   override def getShardLocations(): CompletionStage[ShardLocations] = shardLocations().toJava
+
+  override def updateShardLocations(locations: Map[ShardId, Address]): Future[Done] = {
+    log.debug("updateShardLocations {} for {}", locations, Key)
+    (replicator ? Update(Key, LWWMap.empty[ShardId, String], WriteLocal, None) { existing =>
+      locations.foldLeft(existing) {
+        case (acc, next) => acc.put(self, next._1, next._2.toString)
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/internal/ExternalShardAllocationClientImpl.scala]:
```suggestion
        case (acc, (shardId, address)) => acc.put(self, shardId, address.toString)
```

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/javadsl/ExternalShardAllocationClient.scala [hunk 2]
```diff
@@ -28,10 +28,21 @@ trait ExternalShardAllocationClient {
...
    * @param shard    The shard identifier
    * @param location Location (akka node) to allocate the shard to
-   * @return Confirmation that the update has been propagated to a majority of cluster nodes
+   * @return Conformation that the update has been written to the local node
    */
   def setShardLocation(shard: ShardId, location: Address): CompletionStage[Done]
 
+  /**
+   * Update all of the provided ShardLocations.
+   * The [[Address]] shoul match one of the nodes in the cluster. If the node has not joined
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/javadsl/ExternalShardAllocationClient.scala]:
```suggestion
   * The [[Address]] should match one of the nodes in the cluster. If the node has not joined
```

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/javadsl/ExternalShardAllocationClient.scala [hunk 3]
```diff
@@ -28,10 +28,21 @@ trait ExternalShardAllocationClient {
...
+   * the cluster yet it will be moved to that node after the first cluster
+   * sharding rebalance after it does.
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/javadsl/ExternalShardAllocationClient.scala]:
```suggestion
   * sharding rebalance it does.
```

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/javadsl/ExternalShardAllocationClient.scala [hunk 4]
```diff
@@ -28,10 +28,21 @@ trait ExternalShardAllocationClient {
...
+   *
+   * @param locations to update
+   * @return Conformation that the update has been written to the local node
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/javadsl/ExternalShardAllocationClient.scala]:
```suggestion
   * @return Confirmation that the update has been written to the local node
```

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/scaladsl/ExternalShardAllocationClient.scala [hunk 5]
```diff
@@ -24,14 +24,25 @@ trait ExternalShardAllocationClient {
...
    *
    * @param shard The shard identifier
    * @param location Location (akka node) to allocate the shard to
    * @return Confirmation that the update has been propagated to a majority of cluster nodes
    */
   def updateShardLocation(shard: ShardId, location: Address): Future[Done]
 
+  /**
+   * Update all of the provided ShardLocations.
+   * The [[Address]] shoul match one of the nodes in the cluster. If the node has not joined
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/scaladsl/ExternalShardAllocationClient.scala]:
```suggestion
   * The [[Address]] should match one of the nodes in the cluster. If the node has not joined
```

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/scaladsl/ExternalShardAllocationClient.scala [hunk 6]
```diff
@@ -24,14 +24,25 @@ trait ExternalShardAllocationClient {
...
+   * the cluster yet it will be moved to that node after the first cluster
+   * sharding rebalance after it does.
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/scaladsl/ExternalShardAllocationClient.scala]:
```suggestion
   * sharding rebalance it does.
```

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/scaladsl/ExternalShardAllocationClient.scala [hunk 7]
```diff
@@ -24,14 +24,25 @@ trait ExternalShardAllocationClient {
...
+   *
+   * @param locations to update
+   * @return Conformation that the update has been propagates to a majority of cluster nodes
```
Comment (seglo) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/scaladsl/ExternalShardAllocationClient.scala]:
```suggestion
   * @return Confirmation that the update has been propagates to a majority of cluster nodes
```

**Review by seglo — APPROVED** [review]
LGTM. Just some typos and suggestions. Also needs some MiMa exceptions for the new APIs.

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by ennru** [comment]
This fails with a complation warning with Scala 2.13 https://travis-ci.org/github/akka/akka/jobs/671989537#L326

File: akka-cluster-sharding/src/main/mima-filters/2.6.4.backwards.excludes/shard-allocation-client.excludes [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# Add methods to trait not for user extension
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/mima-filters/2.6.4.backwards.excludes/shard-allocation-client.excludes:1]: should now be in 2.6.5 excludes

Comment (chbatey) [reply akka-cluster-sharding/src/main/mima-filters/2.6.4.backwards.excludes/shard-allocation-client.excludes:1]: Doh

**Pull request merged**